Make sure crrExistingObjects does not break OOB object

francoisferrand opened the pull request:
The code used to re-create the replication state, which could break some
fields (esp. dataLocation) or remove some replication backends.

We now simply update the status, which is safer (more future-proof) and
simpler.

In addition, the code will now avoid updating the internal version id (not
needed), or properly setup the "null version" flags if instructed to do this
"upgrade" of the object.

Issue: S3UTILS-77

francoisferrand committed:
Update the replication status directly

The code used to re-create the replication state, which could break some
fields (esp. dataLocation) or remove some replication backends.

We now simply update the status, which is safer (more future-proof) and
simpler.

Issue: S3UTILS-77

bert-e commented:
# Hello francoisferrand,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/s3utils/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue S3UTILS-77 contains:


* *None*



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `1.13.1`


Please check the `Fix Version/s` of S3UTILS-77, or the target
branch of this pull request.

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers

jonathan-gramain commented on crrExistingObjects.js:
```diff
@@ -149,26 +149,9 @@ function _markObjectPending(
-            };
-            objMD.setReplicationInfo(replicationInfo);
+
+            objMD.setReplicationSiteStatus(storageClass, TARGET_REPLICATION_STATUS);
```
I think we want to set `'PENDING'` here, not `TARGET_REPLICATION_STATUS`: the target is what objects are targeted for the update, not what status we want to put in metadata.

Also, I don't know if that would work in all cases: for example, new objects don't have any replication status in their metadata, and it's an important use case of the script to be able to replicate new objects without any replication config attached to their bucket when they were created.

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers






The following reviewers are expecting changes from the author, or must review again:

* @jonathan-gramain

francoisferrand committed:
Initialize replicationInfo if not present

This is especially needed in case the object was created before
replication is enabled on the bucket.

Issue: S3UTILS-76

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers

francoisferrand committed:
Fix handling of null-version-id objects

When an object does not have a null-version-id, the "fix" was not
correct, as we did not set all fields (`isNull` and `nullVersionId`).
This is detailed in [1].

Upgrading is not actually needed to set the REPLICATION status, so the
default is now to skip this format change, and just update the metadata
in the object.

Alternatively, it is also possible to set GENERATE_INTERNAL_VERSION_ID
env variable so that these objects actually get migrated to the new
"versionned" format, with an allocated internal version id, but while
keeping the external "null" version id.

[1] https://github.com/scality/cloudserver/blob/development/8.5/docs/ARCHITECTURE.rst#null-version-management

Issue: S3UTILS-77

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue S3UTILS-77 contains:


* `1.13.1`



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `1.13.2`


Please check the `Fix Version/s` of S3UTILS-77, or the target
branch of this pull request.

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers

welansari commented:
@francoisferrand tested in ARTESCA 1.4.2 on a bucket that has both OOB and replication enabled, and that contain objects with a FAILED replication status.
The S3C bucket used for OOB contains zero byte objects, non MPU normal objects and MPU objects. All objects have been tested in their versioned and non-versioned format.
The command used is the following:
```
sudo kubectl --kubeconfig=/etc/kubernetes/admin.conf run -i --tty --rm s3utils --image=registry.scality.com/s3utils-dev/s3utils:6ea6808c548f55fa7dffdfa2140fe76080efc664 --env="ACCESS_KEY=<ACCESS_KEY>" \
--env "SECRET_KEY=<SECRET_KEY>" \
--env "ENDPOINT=artesca-data-connector-s3api.zenko.svc.cluster.local:8000" \
--env "MONGODB_REPLICASET=data-db-mongodb-sharded-mongos-0.data-db-mongodb-sharded.zenko.svc:27017" \
--env "MONGODB_AUTH_USERNAME=<MONGO_USERNAME>" \
--env "MONGODB_AUTH_PASSWORD=<MONGO_PASSWORD>" \
--env "MONGODB_SHARD_COLLECTIONS=true" \
--env "MONGODB_DATABASE=<MONGO_DB>" \
--env "REPLICATION_GROUP_ID=RG001" \
--env "STORAGE_TYPE=aws_s3" \
--env "TARGET_REPLICATION_STATUS=FAILED" \
--env "SITE_NAME=aws-walid-test" \
--restart=Never -n zenko -- node crrExistingObjects.js another-bucket
```
Replication restarted for all of the failed objects, the objects got replicated successfully and are still accessible.

erwan-b commented:
/bypass_author_approval

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/1.13`



The following branches will **NOT** be impacted:

* `development/1.12`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **bypass_author_approval**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/1.13`



The following branches have **NOT** changed:

* `development/1.12`



Please check the status of the associated issue S3UTILS-77.

Goodbye francoisferrand.
